mlir/lib/Dialect/MIGraphX/IR/MIGraphX.cpp

```diff
@@ -247,29 +247,29 @@ void MIXRShapedType::getStridesToStandardShapePermutation(
   ArrayRef<int64_t> strides = getStrides();
   size_t n = strides.size();
   ret.resize_for_overwrite(n);
-  // Vector that'll be sorted so that the dimension with the smallest stride
-  // comes first. The second parameter is the reversed index of the dimension
-  // in the original strides array, which ensures that sorting by < leads
-  // to shapes like <1x2x1x3xf32, 6x3x3x1> are not permuted.
   SmallVector<std::pair<int64_t, size_t>, 4> indexedStrides;
   for (auto [idx, stride] : llvm::enumerate(strides)) {
     if (stride == 0) {
       // Broadcast dimensions stay put.
       ret[idx] = idx;
     } else {
-      indexedStrides.push_back({stride, n - idx});
+      indexedStrides.push_back({stride, idx});
```

Comment:
Using `idx` or `n - idx` wont work correct for many cases.

Using `n - idx`, then cases like `reorder_shape({shape::float_type, {1, 24, 24, 1}}, {0, 3, 1, 2})` will have a permutation of `{0, 2, 1, 3}` instead of `{0, 3, 1, 2}`.

Using `idx`, is even less correct as cases with standard shape like `shape{shape::float_type, {64, 1, 24, 24}}` gives a permutation of `{1, 0, 2, 3}` instead of `{0, 1, 2, 3}`.

The algorithm that is used by both [migraphx](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/permutation.cpp#L44) and [miopen](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/include/miopen/tensor.hpp#L218) is to sort with the strides and dimensions(or `lens`):

```cpp
std::vector<int64_t> find_permutation(const shape &s) {
  std::vector<std::int64_t> result(s.lens().size());
  std::iota(result.begin(), result.end(), 0);
  std::stable_sort(result.begin(), result.end(), [&](auto x, auto y) {
    return std::make_tuple(s.strides()[x], s.lens()[x]) >
           std::make_tuple(s.strides()[y], s.lens()[y]);
  });
  return result;
}
```

This came from a long discusson with me and @jerryyin on how to handle many of these edge cases(see discussion [here](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/765#discussion_r596465551)). There are a lot of unit tests in migraphx to check the correctness of this implementation as well, so I strongly recommend using this well tested version over the version here that doesnt even have unit tests.

Reply:
Ok, hold on, that doesn't make sense. How is the length of the dimension being sorted the tiebreaker? Wouldn't it be more reasonable to break ties between strides by position in the permuted shape?

Also, the permutations you posted are inverses of each other, so I wonder if some of the confusion here is due to us trying to generate two different permutation.

The function I'm writing is the one, that given a shape `s`, produces a permutation `p` such that `p(strides(s))` is in standard order and that is the identity on broadcasted dimensions

Reply:
>  How is the length of the dimension being sorted the tiebreaker? 

Because the 1 dimensions repeats the strides of the previous dimensions. So when strides are equal you want the 1 dimension to come first and then the dimension that is not 1 to come after.

> Wouldn't it be more reasonable to break ties between strides by position in the permuted shape?

No because the shape has already been permutated so you dont know which dimension originally came first. 

> The function I'm writing is the one, that given a shape s, produces a permutation p such that p(strides(s)) is in standard order

Yes that is the inverse(Usually people refer to the permutation as we wrote it in migraphx(ie NHWC or 0231))

We compute the inverse permutation by using a sort with `std::less`(see [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/permutation.cpp#L39)), so it might be possible to get `find_permutation` to produce this permutation by flipping the `>` to a `<` instead in the sort comparator.

Reply:
I'll move to your logic and find a good name for it

Reply:
> We compute the inverse permutation by using a sort with std::less(see [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/permutation.cpp#L39)), so it might be possible to get find_permutation to produce this permutation by flipping the > to a < instead in the sort comparator.

So this probably wont work, I think it just reverses the permutation instead of inverting it. Maybe you could do something like `std::make_tuple(s.strides()[x], std::numeric_limits<int64_t>::max() - s.lens()[x])` in the comparator, but I think just inverting it would be simpler.